src/io/simulator/simulator_handle.hpp

```diff
@@ -0,0 +1,453 @@
+  std::function<bool(void *)> is_awaited_;
+  std::function<void(void *, OpaqueMessage)> fill_;
+  std::function<void(void *)> time_out_;
+
+ public:
+  OpaquePromise(OpaquePromise &&old)
+      : ti_(old.ti_),
+        ptr_(old.ptr_),
+        dtor_(old.dtor_),
+        is_awaited_(old.is_awaited_),
+        fill_(old.fill_),
+        time_out_(old.time_out_) {
+    old.ptr_ = nullptr;
+  }
+
+  OpaquePromise &operator=(OpaquePromise &&old) {
+    MG_ASSERT(this != &old);
+
+    ptr_ = old.ptr_;
+    ti_ = old.ti_;
+    dtor_ = old.dtor_;
+    is_awaited_ = old.is_awaited_;
+    fill_ = old.fill_;
+    time_out_ = old.time_out_;
+
+    old.ptr_ = nullptr;
+
+    return *this;
+  }
+
+  OpaquePromise(const OpaquePromise &) = delete;
+  OpaquePromise &operator=(const OpaquePromise &) = delete;
+
+  template <typename T>
+  std::unique_ptr<ResponsePromise<T>> Take() {
+    MG_ASSERT(typeid(T) == *ti_);
+    MG_ASSERT(ptr_ != nullptr);
+
+    ResponsePromise<T> *ptr = static_cast<ResponsePromise<T> *>(ptr_);
+
+    ptr_ = nullptr;
+
+    return std::unique_ptr<T>(ptr);
+  }
+
+  template <typename T>
+  explicit OpaquePromise(std::unique_ptr<ResponsePromise<T>> promise)
+      : ti_(&typeid(T)),
+        ptr_(static_cast<void *>(promise.release())),
+        dtor_([](void *ptr) { static_cast<ResponsePromise<T> *>(ptr)->~ResponsePromise<T>(); }),
+        is_awaited_([](void *ptr) { return static_cast<ResponsePromise<T> *>(ptr)->IsAwaited(); }),
+        fill_([](void *this_ptr, OpaqueMessage opaque_message) {
+          T message = std::any_cast<T>(std::move(opaque_message.message));
+          auto response_envelope = ResponseEnvelope<T>{.message = std::move(message),
+                                                       .request_id = opaque_message.request_id,
+                                                       .from_address = opaque_message.from_address};
+          ResponsePromise<T> *promise = static_cast<ResponsePromise<T> *>(this_ptr);
+          auto unique_promise = std::unique_ptr<ResponsePromise<T>>(promise);
+          unique_promise->Fill(std::move(response_envelope));
+        }),
+        time_out_([](void *ptr) {
+          ResponsePromise<T> *promise = static_cast<ResponsePromise<T> *>(ptr);
+          auto unique_promise = std::unique_ptr<ResponsePromise<T>>(promise);
+          ResponseResult<T> result = TimedOut{};
+          unique_promise->Fill(std::move(result));
+        }) {}
+
+  bool IsAwaited() {
+    MG_ASSERT(ptr_ != nullptr);
+    return is_awaited_(ptr_);
+  }
+
+  void TimeOut() {
+    MG_ASSERT(ptr_ != nullptr);
+    time_out_(ptr_);
+    ptr_ = nullptr;
+  }
+
+  void Fill(OpaqueMessage &&opaque_message) {
+    MG_ASSERT(ptr_ != nullptr);
+    fill_(ptr_, std::move(opaque_message));
+    ptr_ = nullptr;
+  }
+
+  ~OpaquePromise() {
+    if (nullptr != ptr_) {
+      dtor_(ptr_);
+    }
+  }
+};
+
+struct DeadlineAndOpaquePromise {
+  uint64_t deadline;
+  OpaquePromise promise;
+};
+
+class SimulatorHandle {
```

Comment:
Could you please move the non-template member function into a source file? They are quite big.